### DISCUSSION TITLE
Fun histogram tweak

### DISCUSSION THREAD

**Discussion by daviewales** [saulpw/visidata#1807] [Show and tell]
Set the following options in your `~/.visidatarc` for a fun histogram style:

options.disp_histogram = '▇'
options.color_current_row = 'green'

**Before**:

![image](https://user-images.githubusercontent.com/2499066/225786211-cfdd4806-e9bc-45df-bf66-f1ecd960ef91.png)

**After**:

![image](https://user-images.githubusercontent.com/2499066/225786077-8d6ab239-5cd7-410b-b59a-82fcad418a0b.png)

The colour change is necessary because the default `reverse` option for `color_current_row` doesn't look great with the '▇' character.

(Data from https://visidata.org/usage.tsv)

**Comment by saulpw** [comment]
Neat, @daviewales!  Just as a note, you can do:

`FreqTableSheet.options.color_current_row = 'green'`

and that option will only take effect on the given sheet type.

**Reply by frosencrantz** [reply]
That's a fun tweak, @daviewales!

I'll also point you to: https://github.com/saulpw/visidata/issues/1621  which allows the code that generates the histogram to be changed.  It shouldn't be hard to make the histogram a little more fine tuned using the Unicode https://en.wikipedia.org/wiki/Block_Elements.  Not sure if this would interest you.  If you do it, please share.

Also, it seems like it would be possible to generate other graphs like timelines or gantt chart  style data like the example in https://github.com/saulpw/visidata/issues/924 as suggested by @geekscrapy.

**Reply by daviewales** [reply]
Thanks @saulpw and @frosencrantz!
I've kind of been enjoying the variety of having _all_ my current rows green, but limiting it to just the FreqTableSheet probably makes more sense.

I will definitely look into those links @frosencrantz. I was wanting to hijack the Frequency table graph to make a bar graph from aggregation columns, so perhaps I can make some progress there.

**Reply by midichef** [reply]
@daviewales I really like your tweak of bar chart histograms! So much so that I think it should be the default!
@frosencrantz I went ahead and made a version that shows the histogram with fractional-width symbols.

Just add this to your .visidatarc:  [histo-visidatarc.txt](https://github.com/saulpw/visidata/files/11608854/histo-visidatarc.txt) [EDIT:  2023-07-16, I updated it to work on the latest develop branch:
[histo-visidatarc.02.txt](https://github.com/saulpw/visidata/files/12065155/histo-visidatarc.02.txt)
which works with the column width changes that are discussed later in this thread, around June 3 ]
Then set `options.disp_histogram` to a string that's longer than 1 symbol.

Unfortunately, the results aren't good with the symbols I tried. daviewales's original was 7/8ths of full height, which gave 1/8th blank space to separate each bar. The partial-width blocks I use as a demo are full height, so each bar blends into its neighbor below.

If anyone has ideas for Unicode symbols that might look better, please let me know! I tried a few ideas, like half-height blocks, and left them as comments in histo-visidatarc.txt with notes on what I didn't like about them. If you want to test different symbols, here's a helper script that generates data with a good distribution for testing fractional width characters: [histo-fractions.py.txt](https://github.com/saulpw/visidata/files/11608861/histo-fractions.py.txt)

Here's the result using the full-height symbols from Block Elements with widths from 1/8th to 8/8ths:
![histo-eights](https://github.com/saulpw/visidata/assets/67946319/304725fe-1843-4a98-85cd-20e388c275b1)
Here's the result using a half-height block as the base, with the lower-left quadrant symbol to get a half-width:
![histo-half-height](https://github.com/saulpw/visidata/assets/67946319/9051f95b-2860-4541-af74-2541cdc7c901)

**Reply by saulpw** [reply]
I'm game for changing the default of `options.disp_histogram`, but probably not the cursor row color, even if it's just on the Frequency table.  (I have this notion that vd should always be functional with a monochrome terminal like `vt100` like I grew up on).

**Reply by saulpw** [reply]
BTW this thread is really fun and I love the energy towards making vd look slick.

**Reply by daviewales** [reply]
@midichef, that's very cool! I'll have to play with some different characters.

**Reply by frosencrantz** [reply]
@midichef  thank you for sharing! This is wonderful! 

I tried this a little and was really happy with the result.  I was  focused on making the histogram more compact, and not thinking about the styling.  I like how you've explored the styling aspects, this is cool!

I'm still interested in what additional line graphs can be created within a sheet, like the gantt charts @geekscrapy mentions, timelines, or other bar graphs, including stacked bar graphs.

**Reply by frosencrantz** [reply]
I was wondering how to make the histogram use the width of the column instead of `options.disp_histolen`?  It seems like the default width of the column should be `options.disp_histolen`, but if I resize the column, the histogram is resized, but the way it works now is that the graph is fixed width. Making the column thinner just truncates part of the graph.

**Reply by saulpw** [reply]
> I was wondering how to make the histogram use the width of the column instead of options.disp_histolen?

Of course it should work this way, I don't know why it isn't already.  I made it so, and removed the `disp_histolen` option, instead using `default_width*2` (40 instead of 50); always good to remove a superfluous option.  We can add the option back if it's actually important to be able to set it yourself now that the column width does the right thing, or change the default `default_width` to 25, or something else you come up with that's even better :)

**Reply by daviewales** [reply]
> I'm game for changing the default of `options.disp_histogram`, but probably not the cursor row color, even if it's just on the Frequency table. (I have this notion that vd should always be functional with a monochrome terminal like `vt100` like I grew up on).

@saulpw, what if we make it so that only the data is highlighted by the cursor row, not the histogram? This will prevent the bars looking weird when the colours are reversed. (This would be a special-case for histogram sheets)

**Reply by daviewales** [reply]
So there I was, innocently reading through the [Notcurses](https://nick-black.com/htp-notcurses.pdf) book, when I encountered the following paragraph in section 7.6 _Stupid Unicode Tricks_:

> Those readers who remember Code Page 437 might recall 0xFD, a “middle half block”. Given a constant
> background, this can be used with the other two half blocks to increase “vertical resolution” by a factor of 3
> (imagine a tank drawn with upper, middle, and lower half blocks moving on a monochromatic background).
> Well, that character was actually “Solid Square/Histogram/Square Bullet” in the CP437 definitions[31], and
> lives on as U+25A0 BLACK SQUARE (■). Despite its name, it will take on the foreground color.

I was already thinking of the visidata histogram, and when I read the words "middle half block", I had to try it.
Here is the result:

**Before**:
<img width="652" alt="Screen Shot 2023-07-05 at 10 56 58 pm" src="https://github.com/saulpw/visidata/assets/2499066/82d10218-515d-4a29-a576-5f609aeffd52">

**After**:
<img width="656" alt="Screen Shot 2023-07-05 at 10 29 29 pm" src="https://github.com/saulpw/visidata/assets/2499066/986a82cc-0e88-417f-873d-972de5d1bb52">

It's not _perfect_, because the block is not _quite_ in the vertical middle, and there's a small gap between each character.
But I think it's pretty close to the best we can get for default visidata, because there's vertical space both above and below which avoids problems with the reverse video line highlight.
And it's a solid improvement over `*`.

@saulpw, would you consider making this the default character for `options.disp_histogram`?
(Bonus points if we can squeeze it into 2.11.1!)
I sent through a PR: #1949

(I think I still prefer the lower-7/8-height block with green highlight for my personal use, not least because it adds a splash of colour!)

**Reply by anjakefala** [reply]
> So there I was, innocently reading through the [Notcurses](https://nick-black.com/htp-notcurses.pdf) book, when I encountered the following paragraph in section 7.6 Stupid Unicode Tricks:

I am not going to be the person making a decision, but I wanted to say that spending time reading a book, hunting for the perfect character, is very good. =)

**Reply by daviewales** [reply]
Further inspiration for plot functionality:

- https://github.com/red-data-tools/YouPlot
- https://github.com/Martin-Nyaga/termplot
- https://github.com/JuliaPlots/UnicodePlots.jl

**Reply by saulpw** [reply]
I've been appreciating the new histogram display, @daviewales!  It's amazing what a one-character change can do.

**Reply by daviewales** [reply]
Glad you like it! 😀
